### PR TITLE
Adding CloudEvents Extensions to PullSubscription.

### PR DESCRIPTION
**Pull request by n3wscott**
This implements the same logic as Decorators but for PullSubscription.

### CODE REVIEW

File: pkg/pubsub/adapter/adapter.go [hunk 1]
```diff
@@ -54,6 +55,14 @@ type Adapter struct {
 	// subscription to use.
 	Subscription string `envconfig:"PUBSUB_SUBSCRIPTION_ID" required:"true"`
 
+	// ExtensionsBased64 is a based64 encoded josn string of a map of
```
Comment (Harwayne) [inline pkg/pubsub/adapter/adapter.go]:
```suggestion
	// ExtensionsBased64 is a based64 encoded json string of a map of
```

File: test/e2e/lifecycle.go [hunk 2]
```diff
@@ -84,6 +85,39 @@ func TearDown(client *Client) {
...
 }
 
+func (c *Client) LogsFor(namespace, name string, gvr schema.GroupVersionResource) (string, error) {
+	// Get all pods in this namespace.
+	pods, err := c.Kube.Kube.CoreV1().Pods(namespace).List(metav1.ListOptions{})
+	if err != nil {
+		return "", err
+	}
+
+	logs := "" //map[string]string{}
```
Comment (Harwayne) [inline test/e2e/lifecycle.go]: Move down to line 110.

Or use the map and get the logs for every matching Pod.

File: test/e2e/lifecycle.go [hunk 3]
```diff
@@ -84,6 +85,39 @@ func TearDown(client *Client) {
...
+	}
+
+	// Did we find a match like the given name?
+	if pod == nil {
+		return "", fmt.Errorf(`pod for "%s/%s" [%s] not found`, namespace, name, gvr.String())
+	}
+
+	// Collect all the logs from all the containers for this pod.
+	if l, err := c.Kube.Kube.CoreV1().Pods(namespace).GetLogs(pod.Name, &corev1.PodLogOptions{}).DoRaw(); err != nil {
+		logs = err.Error()
```
Comment (Harwayne) [inline test/e2e/lifecycle.go]:
```suggestion
		return "", err
```

File: pkg/apis/pubsub/v1alpha1/pull_subscription_validation.go [hunk 4]
```diff
@@ -139,7 +139,7 @@ func (current *PullSubscription) CheckImmutableFields(ctx context.Context, og ap
 	// Modification of Sink and Transform allowed. Everything else is immutable.
```
[the remark this answers is not in the document]
Comment (n3wscott) [reply pkg/apis/pubsub/v1alpha1/pull_subscription_validation.go]: Fixed.

File: test/e2e/lifecycle.go (hunk 3 above)
Comment (n3wscott) [reply test/e2e/lifecycle.go]: meh, this is in e2e and still working with it.

File: test/e2e/lifecycle.go (hunk 2 above)
Comment (n3wscott) [reply test/e2e/lifecycle.go]: this e2e code is wip.

Comment (Harwayne) [reply test/e2e/lifecycle.go]: OK...is the overall PR WIP or will it be merged with WIP e2e code as-is?

**Comment by n3wscott** [comment]
@Harwayne only the log slurping in e2e is wip

**Comment by Harwayne** [comment]
> @Harwayne only the log slurping in e2e is wip

Then should we remove that code from this PR and submit the rest? I'm hesitant to merge WIP code, including e2e code.

**Pull request merged**